node/src/components/consensus/era_supervisor.rs

```diff
@@ -181,47 +180,32 @@ where
 
-        let bonded_eras = era_supervisor.bonded_eras();
-        let era_ids: Vec<EraId> = era_supervisor
-            .iter_past(current_era, bonded_eras.saturating_mul(3))
-            .collect();
-
         // Collect the information needed to initialize all recent eras.
-        // TODO: Merge these into one
-        let mut key_blocks = BTreeMap::new();
-        let mut booking_blocks = BTreeMap::new();
-        for era_id in era_ids {
-            let minimum_era = era_supervisor
-                .protocol_config
-                .last_emergency_restart
-                .unwrap_or_else(|| EraId::from(0u64));
-            let key_block_era_id = era_id.saturating_sub(1);
-            let block_header = storage
-                .read_switch_block_header_by_era_id(key_block_era_id)?
+        let auction_delay = era_supervisor.protocol_config.auction_delay;
+        let last_activation_point = era_supervisor.protocol_config.last_activation_point;
+        let mut switch_blocks = Vec::new();
+        // We need to initialize current_era, current_era - 1 and current_era - 2. To initialize an
+        // era, all switch blocks between its booking block and its key block are required. The
+        // booking block for era N is in N - auction_delay - 1, and the key block in N - 1.
+        // So we need all switch blocks between (including) current_era - 2 - auction_delay - 1
+        // and (excluding) current_era. However, we never use any block from before the last
+        // activation point.
+        let earliest_era =
+            last_activation_point.max(current_era.saturating_sub(auction_delay).saturating_sub(3));
+        for era_id in (earliest_era.value()..current_era.value()).map(EraId::from) {
+            let switch_block = storage
+                .read_switch_block_header_by_era_id(era_id)?
                 .ok_or_else(|| anyhow::Error::msg("No such switch block"))?;
-            key_blocks.insert(era_id, block_header);
-            let booking_block_era_id = era_id
-                .saturating_sub(era_supervisor.protocol_config.auction_delay)
-                .saturating_sub(1);
-            let booking_block_hash = if booking_block_era_id < minimum_era {
-                Default::default()
-            } else {
-                storage
-                    .read_switch_block_header_by_era_id(booking_block_era_id)?
-                    .ok_or_else(|| anyhow::Error::msg("No such switch block"))?
-                    .hash()
-            };
-            booking_blocks.insert(era_id, booking_block_hash);
+            switch_blocks.push(switch_block);
+        }
+        // The create_new_era method initializes the era that the slice's last block is the key
+        // block for. We want to initialize the three latest eras, so we have to pass in the whole
+        // slice for the current era, and omit one or two elements for the other two. We never
+        // initialize the last_activation_point or an earlier era, however.
```

Comment:
Consider a worked example with `Era(5)`, `Era(4)` and `Era(3)`

Reply:
Done in https://github.com/casper-network/casper-node/pull/2127/commits/429a2b1af6acf011c0a141bc198e9bb52ecb75b8.